python/engine/evaluation.py

```
@@ -356,6 +356,115 @@ def eval_coherence_samples(gpu_model, cfg, max_tokens: int = 30,
+
+    def _kl(p: "np.ndarray", q: "np.ndarray") -> float:
+        eps = 1e-10
+        p = p + eps
+        q = q + eps
+        return float(np.sum(p * np.log(p / q)))
+
+    level_probs: list = []
+    levels_data: list = []
+
+    memory = host_context.memory  # list[k] of flat Vec<f32> length d*d each
+    for l in range(k):
```

Ruff (0.15.2)

[error] Ambiguous variable name: `l`

(E741)